router_macro/src/lib.rs

```diff
@@ -45,32 +41,102 @@ impl SegmentParser {
         for input in self.input.by_ref() {
             match input {
                 TokenTree::Literal(lit) => {
+                    let lit = lit.to_string();
+                    if lit.contains("//") {
+                        abort!(
+                            proc_macro2::Span::call_site(),
+                            "Consecutive '/' is not allowed"
+                        );
+                    }
                     Self::parse_str(
-                        lit.to_string()
-                            .trim_start_matches(['"', '/'])
-                            .trim_end_matches(['"', '/']),
                         &mut self.segments,
+                        lit.trim_start_matches(['"', '/'])
+                            .trim_end_matches(['"', '/']),
                     );
                 }
-                TokenTree::Group(_) => todo!(),
-                TokenTree::Ident(_) => todo!(),
-                TokenTree::Punct(_) => todo!(),
+                TokenTree::Group(_) => unimplemented!(),
+                TokenTree::Ident(_) => unimplemented!(),
+                TokenTree::Punct(_) => unimplemented!(),
+            }
+        }
+    }
+
+    pub fn parse_str(segments: &mut Vec<Segment>, current_str: &str) {
+        if ["", "*"].contains(&current_str) {
+            return;
+        }
+
+        for segment in current_str.split('/') {
+            if let Some(segment) = segment.strip_prefix(':') {
+                segments.push(Segment::Param(segment.to_string()));
+            } else if let Some(segment) = segment.strip_prefix('*') {
+                segments.push(Segment::Wildcard(segment.to_string()));
+            } else {
+                segments.push(Segment::Static(segment.to_string()));
+            }
+        }
+    }
+}
+
+impl Segment {
+    fn is_valid(segment: &str) -> bool {
+        segment
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
```

Comment:
I'm wondering if this is slightly too restrictive — for example, I'd think `_` should be a valid character for static segments, at least, and possibly other [rarer but spec-compliant characters](https://stackoverflow.com/questions/4669692/valid-characters-for-directory-part-of-a-url-for-short-links).

Reply:
Great point!

I've added the following permitted characters:
```rust
const RFC3986_UNRESERVED: [char; 4] = ['-', '.', '_', '~'];
const RFC3986_PCHAR_OTHER: [char; 1] = ['@'];
```

And also added uppercase ascii:
```rust
c.is_ascii_digit()
    || c.is_ascii_lowercase()
    || c.is_ascii_uppercase()
    || RFC3986_UNRESERVED.contains(&c)
    || RFC3986_PCHAR_OTHER.contains(&c)
```

Technically `RFC3986_PCHAR_OTHER` also permits `:`, however I've removed it due to potential clashes with param logic.

Reply:
Thanks! This seems good to me. If someone really `:` in a static route they can always use the tuple syntax... This is just a convenience.